### PR TITLE
Add docstring for saved tensors default hooks

### PR DESCRIPTION
**Pull request by Varal7**
Add documentation for the saved tensors default hooks introduced in #61834 / #62563

Sister PR: #62362 (will add a link from autograd.rst to notes/autograd in whatever PR does not land first)

### CODE REVIEW

File: torch/autograd/saved_variable_default_hooks.py [hunk 1]
```diff
@@ -1,13 +1,85 @@
 import torch
-from typing import Any
+from typing import Callable, Any
 
-def set_saved_tensors_default_hooks(pack_hook, unpack_hook):
+def set_saved_tensors_default_hooks(pack_hook: Callable[[torch.Tensor], Any], unpack_hook: Callable[[Any], torch.Tensor]):
+    r"""Sets a pair of pack / unpack hooks for saved tensors.
+
+    Use these hooks to define how intermediary results of an operation
+    should be packed before saving, and unpacked on retrieval.
```
Comment (Varal7) [inline torch/autograd/saved_variable_default_hooks.py:8]: Here is where I would add a link to https://github.com/pytorch/pytorch/pull/62362 which defines the terms `packing` and `unpacking`.

File: torch/autograd/__init__.py [hunk 2]
```diff
@@ -20,7 +20,7 @@
 from ..overrides import has_torch_function, handle_torch_function
 from . import functional
 from . import forward_ad
-from . import saved_variable_default_hooks as graph
```
Comment (Varal7) [inline torch/autograd/__init__.py:23]: Somehow autodoc can't find the `graph` module if it's imported
with an alias.
https://www.internalfb.com/intern/opensource/ci/job/log/265181241677988/

File: docs/source/autograd.rst [hunk 3]
```diff
@@ -241,3 +241,19 @@ Anomaly detection
...
+Some operations need intermediary results to be saved during the forward pass
+in order to execute the backward pass.
+You can define how these saved tensors should be packed / unpacked using hooks.
+
+.. autosummary::
+    :toctree: generated
+    :nosignatures:
+
+    graph.set_saved_tensors_default_hooks
+    graph.reset_saved_tensors_default_hooks
```
Comment (zou3519) [inline docs/source/autograd.rst:258]: Hmm, I'd argue that these should be context managers especially if we are going to change the default hooks to be thread-local. I'm fine with having a function-only option in the future, but only if we heavily advertise the context manager version (for safety) like we do with torch.set_grad_enabled

File: docs/source/autograd.rst [hunk 4]
```diff
@@ -241,3 +241,15 @@ Anomaly detection
...
 
 .. autoclass:: set_detect_anomaly
+
+
+Saved tensors default hooks
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Some operations need intermediary results to be saved during the forward pass
+in order to execute the backward pass.
+You can define how these saved tensors should be packed / unpacked using hooks.
```
Comment (zou3519) [inline docs/source/autograd.rst:251]: Some more motivation here would be nice. E.g. "you might want to reduce memory if your model is just slightly bigger than the GPU memory you're using", "you might want to swap out to disk", etc.

File: torch/autograd/graph.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    The hooks should have the following signatures:
+
+        pack_hook(tensor: Tensor) -> Any
+
+        unpack_hook(Any) -> Tensor
+
+    where the return value of ``pack_hook`` is a valid input to ``unpack_hook``.
+
+    In general, you want ``unpack_hook(pack_hook(t))`` to be equal to ``t`` in terms
+    of value, size and dtype.
```
Comment (zou3519) [inline torch/autograd/graph.py:34]: what about device?

File: torch/autograd/graph.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        self.unpack_hook = unpack_hook
+
+    def __enter__(self):
+        torch._C._autograd._register_default_hooks(self.pack_hook, self.unpack_hook)
+
+    def __exit__(self, *args: Any):
+        torch._C._autograd._reset_default_hooks()
+
+
+class save_on_cpu():
```
Comment (zou3519) [inline torch/autograd/graph.py:75]: nit: you might be able to implement `save_on_cpu` in terms of `saved_tensors_hooks` but it doesn't save too many LOC

**Review by zou3519 — APPROVED** [review]

**Pull request closed without merging**